internal/uvm/cimfs.go

```diff
@@ -0,0 +1,159 @@
+//go:build windows
+// +build windows
+
+package uvm
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+
+	"github.com/Microsoft/go-winio/pkg/guid"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/protocol/guestrequest"
+	"github.com/Microsoft/hcsshim/internal/protocol/guestresource"
+	"github.com/Microsoft/hcsshim/internal/uvm/scsi"
+	"github.com/Microsoft/hcsshim/pkg/cimfs"
+	"github.com/sirupsen/logrus"
+)
+
+type UVMMountedBlockCIMs struct {
+	scsiMounts []*scsi.Mount
+	// Volume GUID, inside the UVM this is the volume at which the CIMs are mounted
+	volumeGUID guid.GUID
+	// SCSI mounts are already ref counted, we only need to ref count the mounted merged CIMs
+	refCount uint32
+	// UVM in which these mounts are done
+	host *UtilityVM
+	// reference key used for ref counting
+	refKey string
+}
+
+func (umb *UVMMountedBlockCIMs) MountedVolumePath() string {
+	return fmt.Sprintf(cimfs.VolumePathFormat, umb.volumeGUID)
+}
+
+func (umb *UVMMountedBlockCIMs) Release(ctx context.Context) error {
+	umb.host.blockCIMMountLock.Lock()
+	defer umb.host.blockCIMMountLock.Unlock()
+
+	if umb.refCount > 1 {
+		umb.refCount--
+		return nil
+	}
+
+	guestReq := guestrequest.ModificationRequest{
+		ResourceType: guestresource.ResourceTypeWCOWBlockCims,
+		RequestType:  guestrequest.RequestTypeRemove,
+		Settings: &guestresource.WCOWBlockCIMMounts{
+			VolumeGUID: umb.volumeGUID,
+		},
+	}
+	if err := umb.host.GuestRequest(ctx, guestReq); err != nil {
+		return fmt.Errorf("failed to mount the cim: %w", err)
+	}
+
+	for i := len(umb.scsiMounts) - 1; i >= 0; i-- {
+		if err := umb.scsiMounts[i].Release(ctx); err != nil {
+			return err
+		}
+	}
+
+	// Remove from cache when ref count reaches zero
+	delete(umb.host.blockCIMMounts, umb.refKey)
+	return nil
+}
+
+// mergedCIM can be nil,
+// sourceCIMs MUST be in the top to bottom order
+func (uvm *UtilityVM) MountBlockCIMs(ctx context.Context, mergedCIM *cimfs.BlockCIM, sourceCIMs []*cimfs.BlockCIM, containerID string) (_ *UVMMountedBlockCIMs, retErr error) {
```

Comment:
is `containerID` used/needed?

Reply:
@ambarve - my change depends on this https://github.com/microsoft/hcsshim/pull/2508/commits/0f887a7ac384e45b8ba30e6b24435a26dd3bfd59#diff-ff14046169f3aa7708334f20575deacb83cb1a6717cb4555c85e6a0c2013eba9R97 - Please add it back. 

Given my PR is dependent on this and will follow right after this, it makes sense to keep this ContainerID here and not let my changes break. @helsaawy Let me know if this is not ok.